databroker/src/broker.rs

```diff
@@ -1561,6 +1623,290 @@ impl<'a, 'b> AuthorizedAccess<'a, 'b> {
+                    let result_can_write_actuator =
+                        self.permissions.can_write_actuator_target(&vss_path);
+                    match result_can_write_actuator {
+                        Ok(_) => {}
+                        Err(PermissionError::Denied) => {
+                            let message = format!("Permission denied for vss_path {}", vss_path);
+                            return Err((ActuationError::PermissionDenied, message));
+                        }
+                        Err(PermissionError::Expired) => {
+                            return Err((
+                                ActuationError::PermissionExpired,
+                                "Permission expired".to_string(),
+                            ))
+                        }
+                    }
+                }
+                Err(ReadError::NotFound) => {
+                    let message = format!("Could not resolve vss_path of vss_id {}", vss_id);
+                    return Err((ActuationError::NotFound, message));
+                }
+                Err(ReadError::PermissionDenied) => {
+                    let message = format!("Permission denied for vss_id {}", vss_id);
+                    return Err((ActuationError::PermissionDenied, message));
+                }
+                Err(ReadError::PermissionExpired) => {
+                    return Err((
+                        ActuationError::PermissionExpired,
+                        "Permission expired".to_string(),
+                    ))
+                }
+            }
+        }
+
+        let provided_vss_ids: Vec<i32> = self
+            .broker
+            .subscriptions
+            .read()
+            .await
+            .actuation_subscriptions
+            .iter()
+            .flat_map(|subscription| subscription.vss_ids.clone())
+            .collect();
+        let intersection: Vec<_> = vss_ids
+            .iter()
+            .filter(|&x| provided_vss_ids.contains(x))
+            .collect();
+        if !intersection.is_empty() {
+            let message = format!(
+                "actuation providers for the following vss_ids already registered: {:?}",
+                intersection
+            );
+            return Err((ActuationError::ProviderAlreadyExists, message));
+        }
+
+        let actuation_subscription: ActuationSubscription = ActuationSubscription {
+            vss_ids,
+            actuation_provider,
+            permissions: self.permissions.clone(),
+        };
+        self.broker
+            .subscriptions
+            .write()
+            .await
+            .add_actuation_subscription(actuation_subscription);
+
+        Ok(())
+    }
+
+    async fn map_actuation_changes_by_vss_id(
+        &self,
+        actuation_changes: Vec<ActuationChange>,
+    ) -> HashMap<i32, Vec<ActuationChange>> {
+        let mut actuation_changes_per_vss_id: HashMap<i32, Vec<ActuationChange>> = HashMap::new();
```

Comment:
```suggestion
        let mut actuation_changes_per_vss_id: HashMap<i32, Vec<ActuationChange>> = HashMap::HashMap::with_capacity(len(actuation_changes));
```
Since we know how big the HashMap could be only reserve the capacity we need.